### PR TITLE
Add FID metric

### LINKED ISSUE #65: [New metric] FID
## 🚀 Feature

Add Fréchet inception distance (FID) metric.
Standard measure for image quality of generative models. 

Originally proposed here:
https://arxiv.org/abs/1706.08500

Other links:
https://en.wikipedia.org/wiki/Fr%C3%A9chet_inception_distance
https://machinelearningmastery.com/how-to-implement-the-frechet-inception-distance-fid-from-scratch/

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Redo of #90 
Fixes #65 
Implements Fréchet inception distance (FID) which is used to asses the quality of images generated with GANs.

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Added FID metric ([]())
```
Comment (SkafteNicki) [inline CHANGELOG.md:12]:
```suggestion
- Added FID metric ([#213](https://github.com/PyTorchLightning/metrics/pull/213))
```

File: setup.py [hunk 2]
```diff
@@ -72,4 +72,7 @@ def _load_py_module(fname, pkg="torchmetrics"):
         'Programming Language :: Python :: 3.8',
         'Programming Language :: Python :: 3.9',
     ],
+    extra_requires={
+        'image-quality': ["torch-fidelity", "scipy"]
```
Comment (Borda) [inline setup.py:76]: shall we just call it an image or vision?

File: requirements/test.txt [hunk 3]
```diff
@@ -17,3 +17,5 @@ cloudpickle>=1.3
 scikit-learn>0.22.1
 scikit-image>0.17.1
 nltk>=3.6
+scipy
```
Comment (Borda) [inline requirements/test.txt:20]: this is not only test, but all so all vision based

File: tests/helpers/datasets.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+import time
+import urllib.request
+from typing import Optional, Sequence, Tuple
+
+import torch
+from torch import Tensor
+from torch.utils.data import Dataset
+
+
+class MNIST(Dataset):
```
Comment (Borda) [inline tests/helpers/datasets.py:26]: this is copy-paste from PL, correct?
so no need to read the details...

File: tests/image_quality/test_fid.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    def __getitem__(self, idx):
+        return self.imgs[idx]
+
+    def __len__(self):
+        return self.imgs.shape[0]
+
+
+@pytest.mark.skipif(
+    not (torch.cuda.is_available() and torch.cuda.device_count() >= 1), reason='test is too slow without gpu'
```
Comment (Borda) [inline tests/image_quality/test_fid.py:107]:
```suggestion
    not torch.cuda.is_available(), reason='test is too slow without gpu'
```
any reason for checking nb divides?

File: requirements/test.txt (hunk 3 above)
Comment (maximsch2) [reply requirements/test.txt:20]: scipy seems to be only used in tests for now, and torch-fidelity is optional for core library?

File: tests/helpers/datasets.py (hunk 4 above)
Comment (SkafteNicki) [reply tests/helpers/datasets.py:26]: yes just a copy

File: tests/image_quality/test_fid.py (hunk 5 above)
Comment (SkafteNicki) [reply tests/image_quality/test_fid.py:107]: nothing particular :]

**Review by edgarriba — COMMENTED** [review]
@SkafteNicki @Borda there is also `pytorch-fid` that implements frechet distance with way more stars in github than torch-fidelity

**Comment by SkafteNicki** [comment]
@edgarriba agree that they have more stars, but they directly state in their readme that:
>>> So if you report FID scores in your paper, and you want them to be exactly comparable to FID scores reported in other papers, you should consider using the official Tensorflow implementation.

`torch-fidelity` on the other hand actually states how close we are on the reference
>>>    Interpolation: 1e-5
    Convolution: 1e-7
    InceptionV3 RGB 8bpp to pool3 features: 1e-4
    Inception Score: 1e-3
    Fréchet Inception Distance: 1e-6
    Kernel Inception Distance: 1e-6

IMO we cannot use `pytorch-fid` because users need to be able to trust that our implementation is correct.

File: tests/helpers/datasets.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            downloaded again.
+    Examples:
+        >>> dataset = MNIST(".", download=True)
+        >>> len(dataset)
+        60000
+        >>> torch.bincount(dataset.targets)
+        tensor([5923, 6742, 5958, 6131, 5842, 5421, 5918, 6265, 5851, 5949])
+    """
+
+    RESOURCES = (
```
Comment (maximsch2) [inline tests/helpers/datasets.py:49]: The files are not that large, maybe let's just commit them to the repo? Especially considering we actually use 1000 images, so the total size will be about 1M if we "inline" the code that generates TrialMNIST. Having test data inlined greatly simplifies testing for cases where test machines don't have internet access (good for security).

We can keep the script the downloads and generates the data here though for reproducibility

Comment (SkafteNicki) [reply tests/helpers/datasets.py:53]: @Borda what do you think?

Comment (Borda) [reply tests/helpers/datasets.py:53]: well if it is just for testing we can make them as LFS to reduce repo size...

**Comment by SkafteNicki** [comment]
@Borda any idea why some builds are failing on the `update enviroment` step?

File: tests/helpers/datasets.py (hunk 6 above)
Comment (maximsch2) [reply tests/helpers/datasets.py:53]: I think this will be pretty small file - under 1 MB IIRC (if we only take TrialMNIST and save preprocessed version that we are actually using) which might not justify LFS, putting it into repo might be just fine?

Comment (Borda) [reply tests/helpers/datasets.py:53]: yes, but then why not have a synthetic dataset rather?

**Comment by Borda** [comment]
it seems to install a different Conda than we requested, you can see that for 1.3 it gets Conda 1.8.1

File: requirements/devel.txt [hunk 7]
```diff
@@ -6,3 +6,6 @@
 
 # add the integration dependencies
 -r integrate.txt
+
+# add extra requirements
+-r image.txt
```
Comment (Borda) [inline requirements/devel.txt:11]:
```suggestion
```
it is already in test.txt

File: tests/helpers/datasets.py (hunk 6 above)
Comment (Borda) [reply tests/helpers/datasets.py:53]: @SkafteNicki ^^ ?

**Review by edgarriba — APPROVED** [review]

File: .github/workflows/ci_test-conda.yml [hunk 8]
```diff
@@ -59,8 +59,9 @@ jobs:
         conda install mkl pytorch=${{ matrix.pytorch-version }} cpuonly
         conda list
         pip --version
-        pip install --requirement requirements.txt --upgrade-strategy only-if-needed --quiet
-        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed --quiet
+        python ./requirements/adjust_versions.py requirements/image.txt
+        pip install --requirement requirements.txt --upgrade-strategy only-if-needed -v
+        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed -v
```
Comment (Borda) [inline .github/workflows/ci_test-conda.yml:64]:
```suggestion
        pip install --requirement requirements.txt --upgrade-strategy only-if-needed --quiet
        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed --quiet
```

File: requirements/adjust_versions.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+import os
+import re
+import sys
+from typing import Dict, Optional
+
+VERSIONS = [
+    dict(torch="1.9.0", torchvision="", torchtext=""),  # nightly
```
Comment (Borda) [inline requirements/adjust_versions.py:7]: not sure what shall be the torch version needed for PT 1.9

File: requirements/adjust_versions.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    dict(torch="1.8.1", torchvision="0.9.1", torchtext="0.9.1"),
+    dict(torch="1.8.0", torchvision="0.9.0", torchtext="0.9.0"),
+    dict(torch="1.7.1", torchvision="0.8.2", torchtext="0.8.1"),
+    dict(torch="1.7.0", torchvision="0.8.1", torchtext="0.8.0"),
+    dict(torch="1.6.0", torchvision="0.7.0", torchtext="0.7"),
+    dict(torch="1.5.1", torchvision="0.6.1", torchtext="0.6"),
+    dict(torch="1.5.0", torchvision="0.6.0", torchtext="0.6"),
+    dict(torch="1.4.0", torchvision="0.5.0", torchtext="0.5"),
```
Comment (Borda) [inline requirements/adjust_versions.py:15]: pls, add also TV version for PT 1.3

File: tests/helpers/datasets.py (hunk 6 above)
Comment (SkafteNicki) [reply tests/helpers/datasets.py]: remove the mnist to just run the test on synthetic dataset as proposed

File: .github/workflows/ci_test-conda.yml [hunk 11]
```diff
@@ -57,10 +56,13 @@ jobs:
...
         conda info
         conda install mkl pytorch=${{ matrix.pytorch-version }} cpuonly
+        conda install cpuonly $(python ./requirements/adjust-versions.py conda)
         conda list
         pip --version
+        python ./requirements/adjust-versions.py requirements.txt
+        python ./requirements/adjust-versions.py requirements/image.txt
         pip install --requirement requirements.txt --upgrade-strategy only-if-needed --quiet
-        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed --quiet
+        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed --quiet --find-links https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
```
Comment (Borda) [inline .github/workflows/ci_test-conda.yml:65]:
```suggestion
        pip install --requirement requirements/test.txt --upgrade-strategy only-if-needed --quiet --find-links https://download.pytorch.org/whl/cpu/torch_stable.html
```

**Review by Borda — APPROVED** [review]
